src/core/Tree/Tree.tsx

```diff
@@ -0,0 +1,216 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+import React from 'react';
+import { CommonProps, useTheme } from '../utils';
+import '@itwin/itwinui-css/css/tree.css';
+import cx from 'classnames';
+
+export const TreeContext = React.createContext<
+  | {
+      nodeDepth: number;
+      subNodeIds?: string[];
+      hasSubNodes: boolean;
+    }
+  | undefined
+>(undefined);
+
+export type NodeData<T> = {
+  subNodes?: Array<T>;
+  nodeId: string;
+  node: T;
+  isExpanded?: boolean;
+  isDisabled?: boolean;
+  isSelected?: boolean;
+  hasSubNodes: boolean;
+};
+
+type FlatNode<T> = {
+  nodeData: NodeData<T>;
+  depth: number;
+  subNodeIds?: Array<string>;
+};
+
+export type TreeProps<T> = {
+  /**
+   * Custom renderer for tree nodes, recommended to use TreeNode component.
+   */
+  nodeRenderer: (props: NodeData<T>) => JSX.Element;
+  /**
+   * Items inside tree.
+   */
+  data: T[];
+  /**
+   * Get the NodeData.
+   */
+  getNode: (node: T) => NodeData<T>;
+} & CommonProps;
+
+/**
+ * Tree
+ * @example
+  <Tree<TreeData>
+    data={data}
+    getNode={getNode}
+    nodeRenderer={(props) => (
+      <TreeNode
+        nodeId={props.nodeId}
+        label={props.node.label}
+        sublabel={props.node.subLabel}
+        onNodeExpanded={onNodeExpanded}
+        onNodeSelected={onSelectedNodeChange}
+        isDisabled={props.isDisabled}
+        isExpanded={props.isExpanded}
+        isSelected={props.isSelected}
+        nodeCheckbox={
+          <Checkbox variant='eyeball' disabled={props.isDisabled} />
+        }
+        icon={<SvgPlaceholder />}
+      />
+    )}
+  />
+ */
+
+export const Tree = <T,>(props: TreeProps<T>) => {
+  const { data, className, nodeRenderer, getNode, ...rest } = props;
+  useTheme();
+
+  const treeRef = React.useRef<HTMLUListElement>(null);
+
+  const handleKeyDown = (event: React.KeyboardEvent<HTMLUListElement>) => {
+    const nodes: Array<HTMLElement> = Array.from(
+      treeRef.current?.querySelectorAll('.iui-tree-node:not(.iui-disabled)') ||
```

Comment:
Have you checked key nav in `Menu` component for example? Could you try to rewrite something similar to there to avoid css selectors as much as possible?

Reply:
Updated keyboard navigation to use `getFocusableElements` instead of css selectors